### PR TITLE
Line reading is incorrect, can lead to connection getting stuck forever

### PR DESCRIPTION
**Pull request by ghost**
The line handling is incorrect and can lead to the connection getting stuck forever:

Basically you cannot just read 16KB and wait until all of it arrives, then check for lines since in some interactivity situations you might be required to respond earlier. Therefore, the only safe way to parse IRC lines is to go byte for byte until a line break is reached, OR read a larger amount but with either a timeout or an exact known amount of bytes being present or with non-blocking sockets (neither of which the code appears to be doing),

I also removed the special buffer class from being used as well as always_iterable() because both make the code much less intuitive to read: they just introduce additional in-between abstractions that barely do anything that prevent people from understanding what the code does immediately.

### CODE REVIEW

**Comment by ghost** [comment]
After a bit of pondering, I think my fix will solve the hangs I observed with a single connection, but break multiple connections handled by a single reactor. I will think a bit about it, it probably needs to be solved with non-blocking sockets.

**Comment by ghost** [comment]
I made an amendment which should make it work properly with multiple connections without one blocking the others.

File: irc/client.py [hunk 1]
```diff
@@ -559,27 +559,49 @@ def as_nick(self, name):
...
+        # Process each non-empty line:
+        next_break = self.buffer.find(b"\n")
+        while next_break >= 0:
+            next_line = self.buffer[:next_break]
+            if next_line.endswith("\r"):
+                next_line = next_line[:-1]
+            self.buffer = self.buffer[next_break:]
+            log.debug("FROM SERVER: %s", next_line)
+            decoded_line = None
+            try:
```
Comment (jaraco) [inline irc/client.py:617]: This decoding technique is hard-coded, violating the explicit hooks advertised for customizing the encoding handling, and making it impossible to decode other encodings.

File: irc/client.py [hunk 2]
```diff
@@ -559,27 +559,49 @@ def as_nick(self, name):
...
+                decoded_line = next_line.decode("utf-8")
+            except UnicodeDecodeError:
+                try:
+                    decoded_line = next_line.decode("latin-1")
+                except Exception as e:
+                    decoded_line = next_line.decode("utf-8", "replace")
```
Comment (jaraco) [inline irc/client.py:623]: I'm pretty sure this code is unreachable, as all bytes are valid latin-1.

File: irc/client.py [hunk 3]
```diff
@@ -559,27 +559,49 @@ def as_nick(self, name):
...
-        # process each non-empty line after logging all lines
-        for line in self.buffer:
-            log.debug("FROM SERVER: %s", line)
-            if not line: continue
-            self._process_line(line)
+        # Process each non-empty line:
+        next_break = self.buffer.find(b"\n")
+        while next_break >= 0:
+            next_line = self.buffer[:next_break]
+            if next_line.endswith("\r"):
```
Comment (jaraco) [inline irc/client.py]: One can't call endswith on bytes with a string.

```
$ python -c "b'input'.endswith('\r')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: endswith first arg must be bytes or a tuple of bytes, not str
```

File: irc/client.py [hunk 4]
```diff
@@ -844,8 +848,8 @@ def links(self, remote_server="", server_mask=""):
     def list(self, channels=None, server=""):
         """Send a LIST command."""
         command = "LIST"
-        channels = ",".join(always_iterable(channels))
-        if channels:
+        if channels != None:
```
Comment (jaraco) [inline irc/client.py:873]: If channels is "abc", always_iterable returns ("abc",), but this code will use ("a", "b", "c").

File: irc/client.py [hunk 5]
```diff
@@ -866,7 +870,11 @@ def motd(self, server=""):
     def names(self, channels=None):
         """Send a NAMES command."""
         tmpl = "NAMES {channels}" if channels else "NAMES"
-        channels = ','.join(always_iterable(channels))
+        if channels != None:
+            if isinstance(channels, str) or isintance(channels,
+                    basestring):
```
Comment (jaraco) [inline irc/client.py:897]: basestring is a NameError on Python 3.

Plus, here you see the value of `always_iterable`, providing an elegant, tested replacement for 4-5 lines of code. It's a pattern I've encountered hundreds of times in Python, and others share the sentiment. I'd like to continue to see the use of it.

File: irc/client.py [hunk 6]
```diff
@@ -592,10 +592,12 @@ def process_data(self):
         next_break = self.buffer.find(b"\n")
         while next_break >= 0:
             next_line = self.buffer[:next_break]
-            if next_line.endswith("\r"):
+            self.buffer = self.buffer[next_break+1:]
+            if next_line.endswith(b"\r"):
```
Comment (jaraco) [inline irc/client.py:611]: Aha - here's the correction for b"\r".

**Comment by jaraco** [comment]
> I also removed the special buffer class from being used as well as always_iterable() because both make the code much less intuitive to read: they just introduce additional in-between abstractions that barely do anything that prevent people from understanding what the code does immediately.

In my opinion, `always_iterable` is worth learning. I do recognize that it requires the reader to know the purpose or to look up the implementation or assume from the name what it does. On the other hand, it allows the user to work from an assumption of its purpose and avoid concerns about branching and context in several lines of code. It's a capability elevating function. Arguably the author (me) should put more work into evangelizing the function or getting it incorporated into the stdlib, but for now its use should be preferred over inlining the functionality. Preferably, this type of style change should happen in an independent PR.

> The line handling is incorrect and can lead to the connection getting stuck forever

I've never (or rarely) encountered this behavior, and I've run this library daily in an active environment for 6 years. Could you open a ticket and describe under what circumstances the connection gets stuck? Is it possible that simply changing the read size from `2**14` to `1`, that would correct the issue? Since the issue appears to be affecting you, perhaps the read chunk size should be configurable.

**Comment by ghost** [comment]
I have read up on read() documentation Unix manuals again, and it appears I was wrong :-) it got stuck somewhere else, probably to quakenet simply throttling me. Sorry for making this misguided merge request

**Comment by ghost** [comment]
More background: I assumed read() would hang until the full amount of request bytes is reached (which it doesn't, it terminates once any data arrives), and when I "fixed" this in my fork it was just a coincidence that the irc server allowed me again..

**Pull request closed without merging**